### PR TITLE
libomf: dogfood builtin hash arrays instead of StringTable

### PR DESCRIPTION
**Pull request by WalterBright**
I've been reluctant to dump the custom string hash table in dmd because the builtin AAs require a double lookup. But if the implementation is called directly, bypassing the layers of cruft in object.d, only one lookup is required. Should be dogfooding the builtin AAs.

This tries them out for one platform, if successful, will roll out to the others.

### CODE REVIEW

**Comment by mathias-lang-sociomantic** [comment]
Maybe instead of solving it in the compiler, we could solve it in the runtime ?
Inserting a value without overriding it is a common enough need.

**Comment by WalterBright** [comment]
Ideally, it should be done in the library. But we're building dmd with an older, stable version of the compiler/library.

File: src/libomf.d [hunk 1]
```diff
@@ -20,17 +20,42 @@ import ddmd.root.array;
...
+//version = STRINGTABLE;
+version (STRINGTABLE)
+{
+    import ddmd.root.stringtable;
+}
+else
+{
+    // Use builtin associative arrays
+
+    alias OmfObjSymbol* Value;
```
Comment (JackStouffer) [inline src/libomf.d]: Should use the new alias syntax

**Comment by mathias-lang-sociomantic** [comment]
> Ideally, it should be done in the library. But we're building dmd with an older, stable version of the compiler/library.

Fair point. Is there an associated issue for the druntime enhancement ?

**Comment by MartinNowak** [comment]
No! Replacing a specialized hash table (w/o support for deletion) w/ our non-inlinable C-style binding hash that relies on TypeInfo for hashing and comparison is simply a bad idea.

Also, as usual adding sth. in a single place doesn't tell you anything about the performance impact if used more widely, e.g. for the symbol table. So I'm not sure how you define successful here.

If you want to help us proceeding with better AA support, add your input here https://github.com/dlang/druntime/pull/1282, https://trello.com/c/RDVhfCZR/27-more-core-aa-feedback.

BTW, we wanted to add add `getOrSet` to avoid double lookup at some point, PRs welcome.

**Comment by WalterBright** [comment]
> The Travis CI build could not complete due to an error

An example of what I meant when I said the Travis CI is not reliable enough to be blocking automerges.

**Comment by wilzbach** [comment]
> An example of what I meant when I said the Travis CI is not reliable enough to be blocking automerges.

That's not an issue of Travis per se (It's very reliable in its runs), the issue is due to failing installations of ldc & dmd. @MartinNowak do you have an idea what's happening?

**Comment by WalterBright** [comment]
> BTW, we wanted to add add getOrSet to avoid double lookup at some point, PRs welcome.

There's more to it than that. See line 278.

**Comment by WalterBright** [comment]
> a) actually use the user-visible interface instead of relying on presumed knowledge of the host compiler internals,

The internal interface hasn't changed in 15 years. Besides, it illustrates what does need to happen to improve the user-visibile interface.

> b) it doesn't lead to code duplication (why even leave the old implementations in?).

Makes it easier to benchmark.

**Comment by MartinNowak** [comment]
> The internal interface hasn't changed in 15 years. Besides, it illustrates what does need to happen to improve the user-visibile interface.

The C API will have to go for decent AA performance, right now I couldn't recommend using D's builtin AA for anything performance critical. Can we please close this PR until we made progress on the AA side?

File: src/libomf.d [hunk 2]
```diff
@@ -203,25 +232,83 @@ final class LibOMF : Library
...
+                os.name = strdup(name);         // copy name[] because it is ephemeral
+                os.om = om;
+
+                /* Need to change the key from name[] to os.name[],
+                 * because the former is ephemeral.
+                 * This uses the inside knowledge that the key
+                 * is stored immediatedly before the value.
+                 */
+                auto pkey = cast(Key*)pvalue - 1;
+                *pkey = os.name[0 .. namelen];
```
Comment (MartinNowak) [inline src/libomf.d]: Relying on the layout of entries is completely unacceptable.

**Comment by MartinNowak** [comment]
>> BTW, we wanted to add add getOrSet to avoid double lookup at some point, PRs welcome.

> There's more to it than that. See line 278.

We're unlikely to ever add an API that allows you to change the key.

**Comment by UplinkCoder** [comment]
I agree with martin here.
our bultin-AAs are non-inline-able.
This will most likey have a huge! performance decrease.

**Comment by DmitryOlshansky** [comment]
I've seen this style of direct AA in DMDScript and it caused all manner of portability problems with new versions of compiler please don't.

**Comment by WalterBright** [comment]
I'm going to withdraw this for now.

**Comment by MartinNowak** [comment]
> I'm going to withdraw this for now.

Can we get some momentum behind the core.aa proposal? I'll have some more time from september on and could manage to push it forward, but so far it's stuck as unapproved.

**Comment by WalterBright** [comment]
> core.aa

link?

**Comment by ZombineDev** [comment]
https://github.com/dlang/druntime#1282

**Comment by WalterBright** [comment]
@ZombineDev thank you

**Pull request closed without merging**